manifests/init.pp

```diff
@@ -1,112 +1,693 @@
+  Array $kernelpkgnames                                   = [],
+  Array $protected_packages                               = [],
+  Array $tsflags                                          = [],
+  Array[Stdlib::Absolutepath] $reposdir                   = [],
+  # lint:ignore:empty_string_assignment
+  Variant[Boolean,String] $distroverpkg                   = '',
+  Variant[Enum[''],Boolean] $alwaysprompt                 = '',
+  Variant[Enum[''],Boolean] $assumeyes                    = '',
+  Variant[Enum[''],Boolean] $clean_requirements_on_remove = '',
+  Variant[Enum[''],Boolean] $color                        = '',
+  Variant[Enum[''],Boolean] $diskspacecheck               = '',
+  Variant[Enum[''],Boolean] $enable_group_conditionals    = '',
+  Variant[Enum[''],Boolean] $exactarch                    = true,
+  Variant[Enum[''],Boolean] $gpgcheck                     = false,
+  Variant[Enum[''],Boolean] $groupremove_leaf_only        = '',
+  Variant[Enum[''],Boolean] $history_record               = '',
+  Variant[Enum[''],Boolean] $keepalive                    = '',
+  Variant[Enum[''],Boolean] $keepcache                    = true,
+  Variant[Enum[''],Boolean] $localpkg_gpgcheck            = '',
+  Variant[Enum[''],Boolean] $obsoletes                    = true,
+  Variant[Enum[''],Boolean] $overwrite_groups             = '',
+  Variant[Enum[''],Boolean] $plugins                      = false,
+  Variant[Enum[''],Boolean] $protected_multilib           = '',
+  Variant[Enum[''],Boolean] $repo_gpgcheck                = '',
+  Variant[Enum[''],Boolean] $reset_nice                   = '',
+  Variant[Enum[''],Boolean] $showdupesfromrepos           = '',
+  Variant[Enum[''],Boolean] $skip_broken                  = '',
+  Variant[Enum[''],Boolean] $ssl_check_cert_permissions   = '',
+  Variant[Enum[''],Boolean] $sslverify                    = '',
+  Variant[Enum[''],Boolean] $tolerant                     = false,
+  Variant[Enum[''],Integer[0, 10]] $debuglevel            = 2,
+  Variant[Enum[''],Integer[0, 10]] $errorlevel            = '',
+  Variant[Enum[''],Integer] $bandwidth                    = '',
+  Variant[Enum[''],Integer] $installonly_limit            = '',
+  Variant[Enum[''],Integer] $mirrorlist_expire            = '',
+  Variant[Enum[''],Integer] $recent                       = '',
+  Variant[Enum[''],Integer] $retries                      = '',
+  Variant[Enum[''],Integer] $timeout                      = '',
+  Variant[Enum[''],Stdlib::Absolutepath] $cachedir        = '/var/cache/yum/$basearch/$releasever',
+  Variant[Enum[''],Stdlib::Absolutepath] $installroot     = '',
+  Variant[Enum[''],Stdlib::Absolutepath] $logfile         = '/var/log/yum.log',
+  Variant[Enum[''],Stdlib::Absolutepath] $persistdir      = '',
+  Variant[Enum[''],Stdlib::Absolutepath] $pluginconfpath  = '',
+  Variant[Enum[''],Stdlib::Absolutepath] $pluginpath      = '',
+  Variant[Enum[''],Stdlib::Absolutepath] $sslcacert       = '',
```

Comment:
Why the empty quotes instead of `undef` ?

Reply:
You are not able to set undef/nil from hiera. With an empty string you can override these settings back to their defaults even if you have set it on another hiera level.

Reply:
Actually I try to avoid the strange case of using an undef boolean to trigger a functionality.
A boolean should only be a boolean and nothing else.

Reply:
+1 a boolean should only be a boolean

You can specify `undef` in yaml by using `~`.

```yaml
foo::bar: ~
```

Reply:
For path's I think we want the following

```puppet
Optional[Stdlib::Absolutepath] $path  = undef,
Optional[Stdlib::Absolutepath] $path2 = '/path/to/file',
```



https://puppet.com/docs/puppet/5.3/lang_data_abstract.html#optional

Reply:
The intention was to always use the empty strings as trigger to avoid adding a param to yum.conf.
Otherwise we end up in some times undef and sometimes empty strings as triggers.

Think using one pattern only is more predictible.